pinot-core/src/main/java/org/apache/pinot/core/transport/InstanceRequestHandler.java

```diff
@@ -0,0 +1,122 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.transport;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.SimpleChannelInboundHandler;
+import javax.annotation.Nullable;
+import org.apache.pinot.common.metrics.ServerMeter;
+import org.apache.pinot.common.metrics.ServerMetrics;
+import org.apache.pinot.common.metrics.ServerQueryPhase;
+import org.apache.pinot.common.request.InstanceRequest;
+import org.apache.pinot.common.utils.BytesUtils;
+import org.apache.pinot.core.query.request.ServerQueryRequest;
+import org.apache.pinot.core.query.scheduler.QueryScheduler;
+import org.apache.thrift.TDeserializer;
+import org.apache.thrift.protocol.TCompactProtocol;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class InstanceRequestHandler extends SimpleChannelInboundHandler<ByteBuf> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(InstanceRequestHandler.class);
+  private static final byte[] EMPTY_BYTES = new byte[0];
+
+  // TODO: make it configurable
+  private static final int SLOW_QUERY_LATENCY_THRESHOLD_MS = 100;
+
+  private final TDeserializer _deserializer = new TDeserializer(new TCompactProtocol.Factory());
+  private final QueryScheduler _queryScheduler;
+  private final ServerMetrics _serverMetrics;
+
+  public InstanceRequestHandler(QueryScheduler queryScheduler, ServerMetrics serverMetrics) {
+    _queryScheduler = queryScheduler;
+    _serverMetrics = serverMetrics;
+  }
+
+  @Override
+  protected void channelRead0(ChannelHandlerContext ctx, ByteBuf msg) {
+    long queryArrivalTimeMs = System.currentTimeMillis();
+    _serverMetrics.addMeteredGlobalValue(ServerMeter.QUERIES, 1);
+    int requestSize = msg.readableBytes();
+    _serverMetrics.addMeteredGlobalValue(ServerMeter.NETTY_CONNECTION_BYTES_RECEIVED, requestSize);
+    byte[] requestBytes = new byte[requestSize];
+    msg.readBytes(requestBytes);
+
+    InstanceRequest instanceRequest = new InstanceRequest();
+    try {
+      _deserializer.deserialize(instanceRequest, requestBytes);
```

Comment:
I think we should offload the netty thread asap. Doing the deserialization inside the event loop thread will limit the concurrency/scalability. Limited, quick work should be done inside it. May be we should decouple this and pass over the burden of building the query request to the thread on server that actually processes the request?

Reply:
This is the same behavior as the current code (check ScheduledRequestHandler), but with lower overhead (no need to construct an immediate future).
If we found this become a bottleneck in the future, we can make it async.